FreeRTOS-Plus/Demo/AWS/Ota_Windows_Simulator/Ota_Over_Http_Demo/DemoTasks/OtaOverHttpDemoExample.c

```diff
@@ -0,0 +1,2558 @@
+ *
+ * @return `MQTTSuccess` if connection succeeds, else appropriate error code
+ * from MQTT_Connect.
+ */
+static MQTTStatus_t prvMQTTConnect( bool xCleanSession );
+
+/**
+ * @brief Register OTA callbacks with the subscription manager.
+ *
+ * @param[in] pTopicFilter The topic filter for which a  callback needs to be registered for.
+ * @param[in] topicFilterLength length of the topic filter.
+ *
+ */
+static void prvRegisterOTACallback(const char* pTopicFilter,
+    uint16_t topicFilterLength);
+
+/**
+ * @brief Suspend OTA demo.
+ *
+ * @return   pPASS or pdFAIL.
+ */
+static BaseType_t prvSuspendOTA(void);
+
+/**
+ * @brief Resume OTA demo.
+ *
+ * @return   pPASS or pdFAIL.
+ */
+static BaseType_t prvResumeOTA(void);
+
+/**
+ * @brief Set OTA interfaces.
+ *
+ * @param[in]  pOtaInterfaces pointer to OTA interface structure.
+ *
+ * @return   None.
+ */
+static void setOtaInterfaces(OtaInterfaces_t* pOtaInterfaces);
+
+/**
+ * @brief Structure containing all application allocated buffers used by the OTA agent.
+ * Structure is passed to the OTA agent during initialization.
+ */
+static OtaAppBuffer_t otaBuffer =
+{
+    .pUpdateFilePath    = updateFilePath,
+    .updateFilePathsize = otaexampleMAX_FILE_PATH_SIZE,
+    .pCertFilePath      = certFilePath,
+    .certFilePathSize   = otaexampleMAX_FILE_PATH_SIZE,
+    .pDecodeMemory      = decodeMem,
+    .decodeMemorySize   = ( 1U << otaconfigLOG2_FILE_BLOCK_SIZE ),
+    .pFileBitmap        = bitmap,
+    .fileBitmapSize     = OTA_MAX_BLOCK_BITMAP_SIZE,
+    .pUrl = updateUrl,
+    .urlSize = OTA_MAX_URL_SIZE,
+    .pAuthScheme = authScheme,
+    .authSchemeSize = OTA_MAX_AUTH_SCHEME_SIZE
+};
+
+/**
+ * @brief Structure used for encoding firmware version.
+ */
+const AppVersion32_t appFirmwareVersion =
+{
+    .u.x.major = APP_VERSION_MAJOR,
+    .u.x.minor = APP_VERSION_MINOR,
+    .u.x.build = APP_VERSION_BUILD,
+};
+
+/**
+ * @brief Registry for all  mqtt topic filters to their corresponding callbacks for OTA.
+ */
+static OtaTopicFilterCallback_t otaTopicFilterCallbacks[] =
+{
+    {
+        .pTopicFilter = OTA_JOB_NOTIFY_TOPIC_FILTER,
+        .topicFilterLength = OTA_JOB_NOTIFY_TOPIC_FILTER_LENGTH,
+        .callback = prvMqttJobCallback
+    },
+    {
+        .pTopicFilter = OTA_DATA_STREAM_TOPIC_FILTER,
+        .topicFilterLength = OTA_DATA_STREAM_TOPIC_FILTER_LENGTH,
+        .callback = prvMqttDataCallback
+    },
+    {
+        .pTopicFilter = OTA_DEFAULT_TOPIC_FILTER,
+        .topicFilterLength = OTA_DEFAULT_TOPIC_FILTER_LENGTH,
+        .callback = prvMqttDefaultCallback
+    }
+};
+
+/*-----------------------------------------------------------*/
+
+static void prvOTAEventBufferFree( OtaEventData_t * const pxBuffer )
+{
+    if( xSemaphoreTake( xBufferSemaphore, portMAX_DELAY ) == pdTRUE )
+    {
+        pxBuffer->bufferUsed = false;
+        ( void ) xSemaphoreGive( xBufferSemaphore );
+    }
+    else
+    {
+        LogError( ( "Failed to get buffer semaphore." ) );
```

Comment:
`portMAX_DELAY` means that `xSemaphoreTake` will not return before getting the semaphore. So this `LogError` will never be reached.